apps/client/src/common/hooks/useEventAction.ts

```diff
@@ -316,5 +324,92 @@ export const useEventAction = () => {
     [_reorderEventMutation],
   );
 
-  return { addEvent, updateEvent, deleteEvent, deleteAllEvents, applyDelay, reorderEvent };
+  /**
+   * Calls mutation to swap events
+   * @private
+   */
+  const _swapEvents = useMutation({
+    mutationFn: requestEventSwap,
+    // we optimistically update here
+    onMutate: async (data) => {
+      // cancel ongoing queries
+      await queryClient.cancelQueries(RUNDOWN_TABLE, { exact: true });
+
+      // Snapshot the previous value
+      const previousEvents: OntimeRundown = queryClient.getQueryData(RUNDOWN_TABLE) ?? [];
+
+      const fromEvent = previousEvents.at(data.from);
+      const toEvent = previousEvents.at(data.to);
+
+      if (!fromEvent || !toEvent) {
+        // Return a context with the previous and new events
+        return { previousEvents };
+      }
+
+      if (fromEvent.type === SupportedEvent.Event && toEvent.type === SupportedEvent.Event) {
+        const eventsWithSwap = previousEvents.map((event, i) => {
```

Comment:
> I wonder whether it would be necessary to iterate through the array since we know exactly where the events we want to swap are.

I'm a bit confused right now. In this [comment](https://github.com/cpvalente/ontime/pull/446#discussion_r1286357835), you suggest we should stick to `id`s since it'd simplify the work. But unless the `OntimeRundown` is a `Map`, I don't see how it's possible to use `id`s to look up an object in a constant complexity.

Does my question make sense?

Reply:
absolutely, sorry for not providing enough context.
I think both statements are correct, with nuance

- The feature / general operation should work with IDs, the context menu will save IDs and it is IDs we send to backend
- In this specific implementation we do a few steps

1. Find events at ids (`fromEvent `and `toEvent`)
2. Check if they are the right type
3. iterate through the array to find the events with matching indexes of data.from and data.to

This seems cumbersome and we can remove a few steps:
1. find event indexes of events at ids (`fromIndex `and `toIndex`) using findIndex
3. replace data between `fromIndex`and `toIndex`

I removed step 2 since I think this should be validated before we get this far

Hope this makes sense

Reply:
> I removed step 2 since I think this should be validated before we get this far

When should this get validated? Also, that `if` is necessary if we want autocomplete on the events. Since there are multiple types of events, TS requires me to check if it's the supported ones before I spread (`{ ...toEvent }`) and return. Does that make sense?

Reply:
In the frontend we KNOW that we only allow the feature on events of type Event. So within our UI, it would not be possible for it to have another value

In the backend, validation should move to the controller. The controllers should be in charge of validation before allowing data inside our domain (the services)

Reply:
> In the frontend we KNOW that we only allow the feature on events of type Event. So within our UI, it would not be possible for it to have another value

But the TS compiler doesn't know. I can cast it, but an `if` isn't that huge of a deal, right?

Reply:
no, the `if` is not a problem. The idea is that we should make sure that we don't get that far into it with the wrong data, since by now we would have already sent the mutation request.

If typescript needs an `if` and you think that is best, its cool with me

I would personally have moved the check into the `swapEvents` function (not, not the `_swapEvents` mutation) and cast if needed

Reply:
Okay, do you think it's safe to assume (once I switch to the `id` implementation) that this action will always be triggered on a `OntimeEvent`? If yes, I'll just cast the variables and won't bother with validating. 

Or should validation should still be done before calling `_swapEvents`?